pkg/services/automation/generator.go

```diff
@@ -0,0 +1,494 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	MaxKubernetesResourceNameLength = 63
+
+	WeGOAppIdentifierLabelKey = "wego.weave.works/app-identifier"
+)
+
+type ConfigMode string
+
+type SourceType string
+
+type ResourceKind string
+
+type ResourceRef struct {
+	Kind           ResourceKind
+	Name           string
+	RepositoryPath string
+}
+
+const (
+	ConfigModeClusterOnly  ConfigMode = "clusterOnly"
+	ConfigModeUserRepo     ConfigMode = "userRepo"
+	ConfigModeExternalRepo ConfigMode = "externalRepo"
+
+	ResourceKindApplication    ResourceKind = "Application"
+	ResourceKindSecret         ResourceKind = "Secret"
+	ResourceKindGitRepository  ResourceKind = "GitRepository"
+	ResourceKindHelmRepository ResourceKind = "HelmRepository"
+	ResourceKindKustomization  ResourceKind = "Kustomization"
+	ResourceKindHelmRelease    ResourceKind = "HelmRelease"
+)
+
+type AutomationGenerator interface {
+	GenerateAutomation(ctx context.Context, app models.Application, clusterName string) ([]AutomationManifest, error)
+}
+
+type AutomationGen struct {
+	GitProvider gitproviders.GitProvider
+	Flux        flux.Flux
+	Logger      logger.Logger
+}
+
+var _ AutomationGenerator = &AutomationGen{}
+
+type AutomationManifest struct {
+	Path    string
+	Content []byte
+}
+
+func NewAutomationGenerator(gp gitproviders.GitProvider, flux flux.Flux, logger logger.Logger) AutomationGenerator {
+	return &AutomationGen{
+		GitProvider: gp,
+		Flux:        flux,
+		Logger:      logger,
+	}
+}
+
+func (a *AutomationGen) getAppSecretRef(ctx context.Context, app models.Application, clusterName string) (GeneratedSecretName, error) {
+	if app.SourceType != models.SourceTypeHelm {
+		return a.getSecretRef(ctx, app, app.GitSourceURL, clusterName)
+	}
+
+	return "", nil
+}
+
+func (a *AutomationGen) getSecretRef(ctx context.Context, app models.Application, url gitproviders.RepoURL, clusterName string) (GeneratedSecretName, error) {
+	var secretRef GeneratedSecretName
+
+	visibility, visibilityErr := a.GitProvider.GetRepoVisibility(ctx, url)
+	if visibilityErr != nil {
+		return "", visibilityErr
+	}
+
+	if *visibility != gitprovider.RepositoryVisibilityPublic {
+		secretRef = CreateRepoSecretName(clusterName, url)
+	}
+
+	return secretRef, nil
+}
+
+func (a *AutomationGen) generateAppAutomation(ctx context.Context, app models.Application, clusterName string) ([]AutomationManifest, error) {
+	a.Logger.Generatef("Generating GitOps automation manifests")
+
+	appManifest, err := generateAppYaml(app)
+	if err != nil {
+		return nil, fmt.Errorf("could not create app.yaml for '%s': %w", app.Name, err)
+	}
+
+	appGoatManifest, err := a.generateApplicationGoat(app, clusterName)
+	if err != nil {
+		return nil, fmt.Errorf("could not create GitOps automation for '%s': %w", app.Name, err)
+	}
+
+	a.Logger.Generatef("Generating application spec manifest")
```

Comment:
This log call should be moved to before the call to `generateAppYaml`

Reply:
Yep